hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/read/S3File.java

```diff
@@ -214,7 +223,40 @@ void close(InputStream inputStream) {
       this.s3Objects.remove(inputStream);
     }
 
-    Io.closeIgnoringIoException(inputStream);
-    Io.closeIgnoringIoException(obj);
+    this.futurePool.executeRunnable(new DrainTask(inputStream, obj));
   }
+
+
+  /**
+   * Drain task that is submitted to the future pool.
+   */
+  private static class DrainTask implements Runnable {
+
+    private final InputStream inputStream;
+    private final S3Object obj;
+    private long drained;
+
+    DrainTask(InputStream inputStream, S3Object obj) {
+      this.inputStream = inputStream;
+      this.obj = obj;
+    }
+
+    @Override
+    public void run() {
+      try {
+
+        while(this.inputStream.read() >= 0) {
+          drained++;
+        }
+
+        LOG.debug("Drained stream of {} bytes", drained);
+
+        Io.closeIgnoringIoException(this.inputStream);
+        Io.closeIgnoringIoException(this.obj);
+      } catch (Exception e) {
```

Comment:
if this happens then the readl: raised an exception. the stream MUST be aborted to stop it being returned to the http connection pool, as its connection is probably broken